### PR TITLE
#13049 Refactored RecordTransformer & merged RecordEnricher

### PR DESCRIPTION
**Pull request by deepthi912**
1) Removed RecordEnricher, RecordEnricherPipeline, RecordEnricherValidationConfig(directly using the boolean instead of having a class for this), RecordEnricherFactory and moved the methods into RecordTransformer.
2) Moved some of the methods from RecordEnricherRegistry into RecordTransformer except static block ---> This will help in recording the details of type of ericher into a map while loading.
3) Removed CLPEncodingEnricherFactory and CustomFunctionEnricherFactory classes and merged into respective EncodingEricher classes

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
We should be able to delete the record enricher related interfaces and general handling classes (e.g. factory, pipeline, registry) after merging it into record transformer package. Can you take a look and see if it is possible to further clean up the code?

**Comment by deepthi912** [comment]
Sure, I wanted to do that too. Will look and see where I can improve further

**Comment by deepthi912** [comment]
@Jackie-Jiang From what I observed, RecordEnricherFactory  is implemented by two types of enrichers(clp and function). Pipeline, registery has some set of static methods calling RecordEnricherFactory and other util classes utilizing them. Although RecordEnricher had similar methods as RecordTransformer, I don't see any interface in RecordTransformer serving the purpose of RecordEnricherFactory. I feel it's also ok to leave the implementation as it was previously for better structure.

**Comment by deepthi912** [comment]
I noticed that registery, pipeline have static methods which are used by some of the classes. These static methods are calling RecordEnricherFactory which is again implemented by function and clp Enricher classes. I tried to check if I could squeeze in anywhere but no right spot. Not sure if by just removing RecordEnricher, we can make any big difference in refactoring though.

**Comment by Jackie-Jiang** [comment]
Basically the idea is to re-do #12243 and integrate CLP transform as a `RecordTransformer`

**Comment by deepthi-912** [comment]
@Jackie-Jiang  PR is ready for review.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 1]
```diff
@@ -78,7 +77,7 @@ public class FlinkSegmentWriter implements SegmentWriter {
   private String _outputDirURI;
   private Schema _schema;
   private Set<String> _fieldsToRead;
-  private RecordEnricherPipeline _recordEnricherPipeline;
+  private RecordTransformer _recordEnricherPipeline;
```
Comment (Jackie-Jiang) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:80]: After changing `RecordEnricher` into `RecordTransformer`, we should be able to put the `RecordEnricher` in front of the existing `RecordTransformer`s as part of the default `CompositeTransformer`

File: pinot-segment-local/src/main/java/org/apache/pinot/plugin/record/enricher/RecordEnricherRegistry.java [hunk 2]
```diff
@@ -16,45 +16,29 @@
...
 import java.util.HashMap;
 import java.util.Map;
 import java.util.ServiceLoader;
-import org.apache.pinot.spi.config.table.ingestion.EnrichmentConfig;
+import org.apache.pinot.segment.local.recordtransformer.RecordTransformer;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
 public class RecordEnricherRegistry {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/plugin/record/enricher/RecordEnricherRegistry.java:29]: I think we should be able to remove this registry and follow the existing way of creating `RecordTransformer`. Same for other places where 2 pipelines exist

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java [hunk 3]
```diff
@@ -275,7 +275,7 @@ public void deleteSegmentFile() {
   private final int _partitionGroupId;
   private final PartitionGroupConsumptionStatus _partitionGroupConsumptionStatus;
   final String _clientId;
-  private final RecordEnricherPipeline _recordEnricherPipeline;
+  private final RecordTransformer _recordEnricherPipeline;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:278]: Same here, we should be able to integrate it into the `TransformPipeline`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/RecordTransformer.java [hunk 4]
```diff
@@ -18,15 +18,79 @@
...
 import org.apache.pinot.spi.data.readers.GenericRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * The record transformer which takes a {@link GenericRow} and transform it based on some custom rules.
  */
 public interface RecordTransformer extends Serializable {
+  final boolean GROOVY_DISABLED = false;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/RecordTransformer.java:43]: Suggest keeping the interface simple. We need to add `getInputColumns()` from `RecordEnricher`, other method seems unnecessary. Take a look at `CompositeTransformer` for record transformer handling

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 5]
```diff
@@ -113,6 +112,29 @@ public class SegmentIndexCreationDriverImpl implements SegmentIndexCreationDrive
   private long _totalStatsCollectorTime = 0;
   private boolean _continueOnError;
 
+  public static void persistCreationMeta(File indexDir, long crc, long creationTime)
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:115]: These changes are not related to this PR. I guess you might enabled `Rearrange code` option in IDE during auto reformat. Can you revert these unrelated changes? There are quite some unrelated auto reformat in several files

Comment (deepthi912) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:115]: Yeah. I will look into this formatting issue.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java (hunk 3 above)
Comment (deepthi912) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:278]: Got it. I will take a look into TransformPipeline. Didn't know this class existed. Will modify according to it.

**Comment by Jackie-Jiang** [comment]
Merge #14601 instead
